### ISSUE TITLE
Feature Request: catalog reduction bookkeeping tools

### ISSUE THREAD

**Issue by aphearin** [astropy/astropy#3056]
Astropy currently provides a cache system for keeping track of the URL from which external data was downloaded, as well as its location on the user's local disk. It is not uncommon in astronomy for this external data to be quite large, and to require reduction before it is suitable for day-to-day use. This Issue is a request for astropy to provide bookkeeping tools for managing the data reductions. 

Without such bookkeeping tools, astropy users typically keep their reduced catalogs stored somewhere on their local disks, and have to "manually remember" the URL from which the catalogs are derived, as well as the various cuts that were applied in the reduction. However, it would be better practice for everyone if the reduced catalogs automatically had metadata attached to them containing the original URL, a frozen copy of the exact code used to reduce the original catalog, as well as various notes the astropy user may find helpful. This way, everyone's reduced catalogs are self-expressive. Bonus behavior would be if the catalogs may be over-written by further uses of the astropy data reduction tools, but only after issuing a warning, and/or optionally changing the "version" of the reduced catalog. 

@eteq

**Comment by hamogu** [comment]
This feature request is essentially describing a database and pipelining system. You download data, keep a log of every step done, possibly keep different versions (e.g. re-download and see if the origin catalog has changed, since e.g. a URL that points to the chandra list of observations will give a different result depending on when you download it, since new observations are added all the time).

Since the downloaded datafiles might be arbitrary formats (not all of which support adding metadata to that file itself), that meta-data would probably have to be kept separately, like log files. That means that any interaction with the data has to happen through wrapper functions that also update the log files and history. 

All in all, this type of system is useful in many cases in astronomy, in particular for large, institutional users (think: the HST pipeline, the ESO archive and data reduction pipeline) where you want to document the origin of every piece of data and all the things that happened to it. For example, https://github.com/nasa/ziggy can do that, but at a very complex level.

I feel that this feature request is way too much to fit within astropy. I don't think that you need to go all the way to [ziggy](https://github.com/nasa/ziggy) or similar systems, but I do think that it's out of scope of astropy itself. I think a relatively lightweight system that's not 100% foolproof, but covers the most common situations described above could be built on top of astropy as an affiliated package.

Thus, I suggest to close this feature request as out-of-scope for astropy itself.
